### PR TITLE
[Impeller] Started recycling vkRenderPass and vkFramebuffer

### PR DESCRIPTION
**Pull request by gaaclarke**
*Replace this paragraph with a description of what this PR is changing or adding, and why. Consider including before/after screenshots.*

*List which issues are fixed by this PR. You must list at least one issue.*

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on writing and running engine tests.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I signed the [CLA].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Review by jonahwilliams — COMMENTED** [review]
You can't recycle the objects like this, you need to use the resource manager and wait for the fence waiter to be signal

**Comment by gaaclarke** [comment]
> You can't recycle the objects like this, you need to use the resource manager and wait for the fence waiter to be signal

It's getting tracked in the tracked resources so it won't be reused until the RingBufferItem is deleted (when the fence waiter signals).  There is a bug in it right now which i'm fixing.

File: impeller/renderer/backend/vulkan/render_pass_vk.cc [hunk 1]
```diff
@@ -592,27 +592,42 @@ bool RenderPassVK::OnEncodeCommands(const Context& context) const {
...
+    RenderPassFrameBufferPair pair = RenderPassFrameBufferPair{
+        .render_pass = render_pass,
+        .framebuffer = framebuffer,
+    };
+
+    render_pass_frame_buffer_pair.GetValue() = std::move(pair);
   }
 
-  if (!encoder->Track(framebuffer) || !encoder->Track(render_pass)) {
+  if (!encoder->Track(
```
Comment (gaaclarke) [inline impeller/renderer/backend/vulkan/render_pass_vk.cc:619]: @jonahwilliams right here we are making sure it isn't placed back into the pool until the frame is done.

**Review by jonahwilliams — COMMENTED** [review]
Ahh I see!

**Comment by gaaclarke** [comment]
This isn't quite right yet.  I hadn't noticed that the vkrenderpass and vkframebuffer creation depends on the render target.  So we will need something like a ring buffer per render target (while making sure we clean up old render targets).

**Comment by gaaclarke** [comment]
Closing this.  After further investigation we've found this wasn't the culprit.

**Comment by jonahwilliams** [comment]
There are multiple culprits, here is a run where despite not being the slowest operation, creating the renderpass still took 3ms:

image

**Comment by gaaclarke** [comment]
Current state: Disabling the cache has no validation errors, turning it on leads to validation errors surrounding the resolve texture:
```
```

**Comment by gaaclarke** [comment]
It tried:
1) setting the layout transition for the resolve texture from coloroptimal->present to undefined->present and that gets us past the first frame, but eventually leads to problems.
2) Introducing a barrier that transitions current->general->current for the resolve texture, like we do for the other color attachments.

My theory is that the synchronization that we are doing for the other textures has to happen for the resolve texture.  But what we are doing for the other textures won't work for the resolve texture since it is swapping between coloroptimal->present so we can't knowingly swap it from general back to whatever.  We need to find a new way.  Given that theory though, I'm at a lose to explain why it currently works.

**Comment by gaaclarke** [comment]
Here is the output now with the new logs: https://gist.github.com/gaaclarke/437c0753782601d0f78300977b7f2c46

It seems like the initialLayout for the resolve texture should be color optimal, or there should be a barrier to make it happen.

**Comment by jonahwilliams** [comment]
Okay I feel dumb, the problem seems to be that the framebuffer objects have references to the actual attachments. So if you recycle them and the attachments aren't the same its just straight up invalid - meaning we always have to recreate them. IN theory the RenderPass is only referencing the attachments in the framebuffer by index, so that may work. Still looking

so basically you can only reuse the framebuffer if you have a static layout with a fixed number of objects per frame

**Comment by gaaclarke** [comment]
My thought process was that render passes should just have a layout which considering the rendertargets are equal, should be the same.  Based on the trace statements we can see that the rendertargets are the same so I didn't implement that part of the caching yet.  I had considered framebufferrs will be harder to recycle since they may have links to swapchain images.  But since the errors I was seeing were related to resolve texture image layouts, which is the province of the renderpass and pipeline barriers I didn't spend much time looking into the framebuffer.

However, allocating the framebuffer is the one that I measured as being slow so if we can't easily recycle that, the value of this PR decreases.

**Comment by gaaclarke** [comment]
Closing this PR to clear out the triage process.  I requested that this policy be reviewed since my hope would be that drafts are not reviewed.

**Pull request closed without merging**